### PR TITLE
feat:basic code for ESP server added

### LINKED ISSUE #47: Add http server code to communicate with PSLab over wifi using ESP
Currently, there is no functionality available to communicate with PSLab wirelessly. This issue aims to add server code for wifi communication

### PR DESCRIPTION
**Pull request by neel1998**
Fix #47

### CODE REVIEW

File: ESP_Server/server.ino [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  for (uint8_t i = 0; i < server.args(); i++) {
+    message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
+  }
+  server.send(404, "text/plain", message);
+}
+
+void setup(void) {
+  WiFi.mode(WIFI_STA);
+  WiFi.begin(ssid, password);
+  //Serial.println("");
```
Comment (CloudyPadmal) [inline ESP_Server/server.ino]: Clean up the code

File: ESP_Server/server.ino [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#include <ESP8266WiFi.h>
+#include <WiFiClient.h>
+#include <ESP8266WebServer.h>
+#include <ESP8266mDNS.h>
+
+#ifndef STASSID
+#define STASSID "ssid"
```
Comment (CloudyPadmal) [inline ESP_Server/server.ino]: No need to define same thing multiple times. Set SSID in line 11 here and remove passwords.

File: ESP_Server/server.ino [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  message += server.args();
+  message += "\n";
+  for (uint8_t i = 0; i < server.args(); i++) {
+    message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
+  }
+  server.send(404, "text/plain", message);
+}
+
+void setup(void) {
+  WiFi.mode(WIFI_STA);
```
Comment (CloudyPadmal) [inline ESP_Server/server.ino]: Mode should be an access point. WIFI_AP

**Comment by neel1998** [comment]
@CloudyPadmal  we cant remove password as far as i think. Because the Wifi.softAP() metho requires 2 arguments. Still I am getting the following error while uploading the sketch,
```
warning: espcomm_sync failed
An error occurred while uploading the sketch
error: espcomm_open failed
error: espcomm_upload_mem failed
```
I have esp-12s module,


and I am making following connections

(left side pins are esp pins and right side pins are arduino)
VCC -> 3.3V
GND -> GND
CH-PC -> 3.3V
Tx -> Tx 
Rx -> Rx
GPIO0 -> GND
RST -> GND

**Comment by CloudyPadmal** [comment]
Password is optional. Look at the documentation https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/soft-access-point-class.html#softap

File: ESP_Server/server.ino [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#include <ESP8266WiFi.h>
+#include <WiFiClient.h>
+#include <ESP8266WebServer.h>
+
+const char* ssid = "ssid";
```
Comment (CloudyPadmal) [inline ESP_Server/server.ino]: SSID should be "PSLab_ESP"  or something

Comment (neel1998) [reply ESP_Server/server.ino]: Yes yes. This is just temporary. Will change .

**Comment by CloudyPadmal** [comment]
> warning: espcomm_sync failed
> An error occurred while uploading the sketch
> error: espcomm_open failed
> error: espcomm_upload_mem failed

Have you programmed as ESP before? :thinking:

**Comment by neel1998** [comment]
> > warning: espcomm_sync failed
> > An error occurred while uploading the sketch
> > error: espcomm_open failed
> > error: espcomm_upload_mem failed
> 
> Have you programmed as ESP before?

Not before this  project.

**Comment by neel1998** [comment]
@CloudyPadmal I haven't interacted with electronics too much before ... But I am willing to work on it.

**Comment by CloudyPadmal** [comment]
@neel1998 you've mentioned in your proposal that you'll have all the background studies covered prior to GSoC. Now it's almost end and I don't think this will have any outcome at all with this feature as you're studying about ESPs now. I dunno I'm not sure what to do. Can we finalize this PR right now? WiFi AP sample source codes are all over the internet and this shouldn't be a WIP for days.

**Comment by neel1998** [comment]
@CloudyPadmal I mentioned in my proposal that I will be providing android side support for esp communication. Still i am trying my best to work on firm ware as well. And for basic AP server code i think this should be fine. We can add other endpoints later.

File: ESP_Server/server.ino (hunk 4 above)
Comment (CloudyPadmal) [reply ESP_Server/server.ino]: Where's the change?

**Comment by neel1998** [comment]
Yes I am not on my system right now. Will make the changes by night

**Pull request merged**